src/core/DistMatrix/ElementMatrix/gpu_instantiate.h

```diff
@@ -0,0 +1,36 @@
+// This will do the entire instantiation of GPU matrices for all known
+// types, then undef everything.
+//
+// The following must be #define-d before #include-ing this file:
+//   - COLDIST
+//   - ROWDIST
+//   - PROTO(T)
+//
+// The definitions of these provided macros will be intact at
+// exit. The macros defined in this file will be all be #undef-d
+// before exit.
+
+#define INST_COPY_CTOR(T,U,V,SRCDEV,TGTDEV)                             \
+  template DistMatrix<T, COLDIST, ROWDIST, ELEMENT, TGTDEV>::DistMatrix( \
+    DistMatrix<T, U, V, ELEMENT, SRCDEV> const&)
+#define INST_ASSIGN_OP(T,U,V,SRCDEV,TGTDEV)                     \
+  template DistMatrix<T,COLDIST,ROWDIST,ELEMENT,TGTDEV>&        \
+  DistMatrix<T, COLDIST, ROWDIST, ELEMENT, TGTDEV>::operator=   \
+  (DistMatrix<T, U, V, ELEMENT, SRCDEV> const&)
+
+#define INST_COPY_AND_ASSIGN(T, U, V)                   \
+    INST_COPY_CTOR(T, U, V, Device::CPU, Device::GPU);  \
+    INST_COPY_CTOR(T, U, V, Device::GPU, Device::CPU);  \
+    INST_COPY_CTOR(T, U, V, Device::GPU, Device::GPU);  \
+                                                        \
+    INST_ASSIGN_OP(T, U, V, Device::CPU, Device::GPU);  \
+    INST_ASSIGN_OP(T, U, V, Device::GPU, Device::CPU)
```

Comment:
Don't we also need the GPU-GPU copy-assignment operator?
```suggestion
    INST_ASSIGN_OP(T, U, V, Device::GPU, Device::CPU)
    INST_ASSIGN_OP(T, U, V, Device::GPU, Device::GPU)
```

Reply:
That case is already handled elsewhere. It should not appear in this list.

Reply:
Ok, I didn't see it in `src/core/DistMatrix/ElementMatrix/*.cpp`, but I'll take your word for it.